### PR TITLE
Add LineBreakSegmenter FFI

### PR DESCRIPTION
**Pull request by sffc**
```bash
ffi/diplomat/cpp/examples/segmenter$ make
g++ -std=c++17 test.cpp ../../../../../target/debug/libicu_capi.a -ldl -lpthread -lm -g
./a.out
Segmenting string:
The quick brown fox jumps over the lazy dog.
0....5....0....5....0....5....0....5....0...
Breakpoints: 4 10 16 20 26 31 35 40 44
```

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]

File: ffi/diplomat/wasm/docs/segmenter_line_ffi.rst [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+        Find the next breakpoint. Returns -1 if at the end of the string or if the index is out of range of a 32-bit signed integer.
+
+.. js:class:: ICU4XLineBreakSegmenter
+
+    An ICU4X line-break segmenter, capable of finding breakpoints in strings. See `the Rust docs <https://unicode-org.github.io/icu4x-docs/doc/icu_segmenter/struct.LineBreakSegmenter.html>`__ for more information.
+
+    .. js:staticfunction:: try_new()
+
+        Construct an :js:class:`ICU4XLineBreakSegmenter` from an locale identifier. See `the Rust docs <https://unicode-org.github.io/icu4x-docs/doc/icu_segmenter/struct.LineBreakSegmenter.html#method.from_bytes>`__ for more information.
```
Comment (aethanyc) [inline ffi/diplomat/wasm/docs/segmenter_line_ffi.rst:16]: Nit: Ditto.

**Review by aethanyc — DISMISSED** [review]

File: ffi/diplomat/src/segmenter_line.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    use icu_segmenter::LineBreakIterator;
+    use icu_segmenter::LineBreakSegmenter;
+
+    #[diplomat::opaque]
+    /// An ICU4X line-break segmenter, capable of finding breakpoints in strings.
+    /// See [the Rust docs](https://unicode-org.github.io/icu4x-docs/doc/icu_segmenter/struct.LineBreakSegmenter.html) for more information.
+    pub struct ICU4XLineBreakSegmenter(LineBreakSegmenter);
+
+    #[diplomat::opaque]
+    pub struct ICU4XLineBreakIterator<'a, 'b>(LineBreakIterator<'a, 'b, char>);
```
Comment (Manishearth) [inline ffi/diplomat/src/segmenter_line.rs:19]: I think this should be a single lifetime, covariance will mean that you won't be restricted by it, but it'll help the diplomat lifetime handling when we add it. 

There's really no benefit to maximum flexibility on lifetimes for types that are primarily used by environments without them.

**Comment by makotokato** [comment]
And, could you file a new issue for UTF-16 string  (`u16string_view `) version of `segment_str`?

File: ffi/diplomat/src/segmenter_line.rs (hunk 2 above)
Comment (sffc) [reply ffi/diplomat/src/segmenter_line.rs]: Done

File: ffi/diplomat/wasm/docs/segmenter_line_ffi.rst (hunk 1 above)
Comment (sffc) [reply ffi/diplomat/wasm/docs/segmenter_line_ffi.rst]: Done

File: ffi/diplomat/cpp/docs/source/segmenter_line_ffi.rst [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+        Find the next breakpoint. Returns -1 if at the end of the string or if the index is out of range of a 32-bit signed integer.
+
+.. cpp:class:: ICU4XLineBreakSegmenter
+
+    An ICU4X line-break segmenter, capable of finding breakpoints in strings. See `the Rust docs <https://unicode-org.github.io/icu4x-docs/doc/icu_segmenter/struct.LineBreakSegmenter.html>`__ for more information.
+
+    .. cpp:function:: static diplomat::result<ICU4XLineBreakSegmenter, std::monostate> try_new()
+
+        Construct an :cpp:class:`ICU4XLineBreakSegmenter` from an locale identifier. See `the Rust docs <https://unicode-org.github.io/icu4x-docs/doc/icu_segmenter/struct.LineBreakSegmenter.html#method.from_bytes>`__ for more information.
```
[the remark this answers is not in the document]
Comment (sffc) [reply ffi/diplomat/cpp/docs/source/segmenter_line_ffi.rst]: Done

**Comment by sffc** [comment]
I added `try_new_with_options` as well as UTF-16 and Latin-1, and addressed the other feedback.

**Review by Manishearth — DISMISSED** [review]

**Review by aethanyc — APPROVED** [review]

**Comment by sffc** [comment]
Note that a side-effect of this PR is that compile times for icu_capi will increase because of the build.rs script in segmenter.  The build.rs script should be moved to icu4x-datagen, however, so this should be a temporary problem.

**Comment by sffc** [comment]
We may also want to consider landing this until after 0.5 is released, because we don't plan on releasing the icu_segmenter crate since it is still in experimental.  @Manishearth thoughts?

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
I think that makes sense.

**Review by makotokato — APPROVED** [review]

**Comment by aethanyc** [comment]
Given that 0.5 is released, are we ready to merge this?

**Comment by sffc** [comment]
Yes!  Let's merge.

**Pull request merged**